tests/apichecks/crds_test.go

```diff
@@ -390,3 +397,151 @@ func TestCRDShortNames(t *testing.T) {
 	want := strings.Join(errs, "\n")
 	test.CompareGoldenFile(t, "testdata/exceptions/shortnames.txt", want)
 }
+
+// Run this test with WRITE_GOLDEN_OUTPUT set to update the exceptions list.
+// Note adding to the exceptions list should not be done trivially.
+func TestCRDFieldPresenceInUnstructured(t *testing.T) {
+	crds, err := crdloader.LoadAllCRDs()
+	if err != nil {
+		t.Fatalf("error loading CRDs: %v", err)
+	}
+
+	unstructs := loadUnstructs(t)
+
+	var errs []string
+	for _, crd := range crds {
+		for _, version := range crd.Spec.Versions {
+
+			if version.Name == "v1alpha1" {
+				continue
+			}
+
+			visitCRDVersion(version, func(field *CRDField) {
+				fieldPath := field.FieldPath
+
+				// Only consider fields under `spec`
+				if !strings.HasPrefix(fieldPath, ".spec.") {
+					return
+				}
+
+				// skip the resource id field
+				if strings.HasSuffix(fieldPath, "resourceID") {
+					return
+				}
+
+				// Check for "Ref" fields
+				if strings.HasSuffix(fieldPath, "Ref") {
+					hasExternal := false
+					hasName := false
+
+					// Check for specific related fields
+					for _, obj := range unstructs {
+						if hasField(obj.Object, fieldPath+".external") {
+							hasExternal = true
+						}
+						if hasField(obj.Object, fieldPath+".name") {
+							hasName = true
+						}
+					}
+
+					// Only report an error if neither external nor name is set
+					if !hasExternal && !hasName {
+						errs = append(errs, fmt.Sprintf("[missing_field] crd=%s version=%v: field %q is not set; neither 'external' nor 'name' are set", crd.Name, version.Name, fieldPath))
+					}
+					return
+				}
+
+				// Skip non-terminal fields (fields with children or slices)
+				if field.props != nil {
+					if len(field.props.Properties) > 0 || field.props.Type == "object" {
+						return
+					}
+					if field.props.Type == "array" && field.props.Items != nil {
+						return // Skip the array itself; focus on its elements
+					}
+				}
+
+				// Any XYZRef field was already handled and handling the children will just double count
+				if strings.Contains(fieldPath, "Ref") {
+					return
+				}
+
+				// Check if field exists in any unstructured object
+				missing := true
+				for _, obj := range unstructs {
+					if hasField(obj.Object, fieldPath) {
```

Comment:
Oooh, can we have another linter that checks for fields we hallucinated :-)  (I.e. checks the other way round)  There might be a refactor where we introduce something like `CollectFields(obj *unstructured.Unstructured) sets[string]`

Reply:
Oh interesting so the check is there to make sure that the fields in the list are defined in the CRD, right ?